ethereum/contracts/relayer/coreRelayer/CoreRelayerBase.sol

```diff
@@ -0,0 +1,264 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.19;
+
+import {IWormhole} from "../../interfaces/IWormhole.sol";
+import {IRelayProvider} from "../../interfaces/relayer/IRelayProvider.sol";
+import {toWormholeFormat, min, pay} from "./Utils.sol";
+import {
+  NoDeliveryInProgress,
+  ReentrantDelivery,
+  ForwardRequestFromWrongAddress,
+  RelayProviderDoesNotSupportTargetChain,
+  RelayProviderQuotedBogusAssetPrice,
+  Send,
+  DeliveryInstruction,
+  ExecutionParameters,
+  IWormholeRelayerBase
+} from "../../interfaces/relayer/IWormholeRelayer.sol";
+import {
+  ForwardInstruction,
+  DeliveryTmpState,
+  getDeliveryTmpState,
+  getRegisteredCoreRelayersState
+} from "./CoreRelayerStorage.sol";
+
+abstract contract CoreRelayerBase is IWormholeRelayerBase {
+  //TODO AMO: see https://book.wormhole.com/wormhole/3_coreLayerContracts.html#consistency-levels
+  //  15 is valid choice perhaps, but it feels messy... (seems like it should be 201?)
+  //  Also, these values should definitely not be defined here but should be provided by IWormhole!
+  uint8 internal constant CONSISTENCY_LEVEL_FINALIZED = 15;
+  uint8 internal constant CONSISTENCY_LEVEL_INSTANT = 200;
+
+  IWormhole private immutable wormhole_;
+  uint16 private immutable chainId_;
+
+  constructor(address _wormhole) {
+    wormhole_ = IWormhole(_wormhole);
+    chainId_ = uint16(wormhole_.chainId());
+  }
+
+  function getRegisteredCoreRelayerContract(uint16 chainId) public view returns (bytes32) {
+    return getRegisteredCoreRelayersState().registeredCoreRelayers[chainId];
+  }
+
+  //Our get functions require view instead of pure (despite not actually reading storage) because
+  //  they can't be evaluated at compile time. (https://ethereum.stackexchange.com/a/120630/103366)
+  
+  function getWormhole() internal view returns (IWormhole) {
+    return wormhole_;
+  }
+
+  function getChainId() internal view returns (uint16) {
+    return chainId_;
+  }
+
+  function publishAndPay(
+    uint256 wormholeMessageFee,
+    uint256 maxTransactionFee,
+    uint256 receiverValue,
+    bytes memory message,
+    uint8 consistencyLevel,
+    IRelayProvider relayProvider
+  ) internal returns (uint64 sequence) { 
+    sequence =
+      getWormhole().publishMessage{value: wormholeMessageFee}(0, message, consistencyLevel);
+
+    emit SendEvent(sequence, maxTransactionFee, receiverValue);
+
+    uint256 amount;
+    unchecked {amount = maxTransactionFee + receiverValue;}
+    //TODO AMO: what if pay fails? (i.e. returns false)
+    pay(relayProvider.getRewardAddress(), amount);
+  }
+
+  // ----------------------- delivery transaction temorary storage functions -----------------------
+
+  function startDelivery(address targetAddress) internal {
```

Comment:
As mentioned on the call: Because they form a coherent whole that belongs together (Send and Delivery coordinate around DeliveryTmpState. Really personal preference of whether to keep them as a logical unit together in this spot or split them to the subclasses where the individual functions are actually used.

Reply:
I assume because this file assumes the role of 'setters' and modifies state directly